Add description of processing of authorized-by (#148)
ietf-rats-wg/draft-ietf-rats-corim: draft-ietf-rats-corim.md

## Evidence Collection

> In the evidence collection phase the Verifier communicates with attesters to

Reviewer: ```suggestion
In the Evidence collection phase the Verifier communicates with attesters to
```

### The Accepted Claims Set

At the end of the Evidence collection process Evidence has been converted into
a format suitable for appraisal. To this end, this document describes an
`accepted-claims-set` format and the algorithms used to compare it against
CoMID Reference Values.

~~~ cddl
{::include cddl/accepted-claims-set.cddl}
~~~

Verifiers are not required to use this as their internal state, for the
purposes of this document a sample Verifier is discussed which uses this format.

The Accepted Claims Set (ACS) contains the actual state of Target Environments (TEs).

> The `evidence-triples` field contains Evidence (from Attestors) and Endorsements

Reviewer 2: ```suggestion
The `evidence-triples` field contains Evidence (from Attesters) and Endorsements
```

> The `evidence-triples` field contains Evidence (from Attesters) and Endorsements

Author: I have changed it to state-triples. I think actual-state-triples is too long.

> The `evidence-triples` field contains Evidence (from Attesters) and Endorsements

Reviewer 2: #158

Reviewer 2: There is some work to be done to fulfil the explanation which will be covered by issue #158

(e.g. from `endorsed-triple-record`).

CoMID Reference Values will be matched against the Accepted Claims Set, as per
the appraisal policy of the Verifier.
This document describes an example evidence structure which can be easily
matched against these Reference Values.

Each entry within `evidence-triples` uses the syntax of `endorsed-triple-record`.
When an `endorsed-triple-record` appears within `evidence-triples` it
indicates that the authority named by `measurement-map`/`authorized-by`
asserts that the actual state of the Target Environment `environment-map`

> has the values in `measurement-map`/`mval`.

Author: Reworded

In `authorized-by`, authority is represented by cryptographic keys. Authority
is asserted by digitally signing a Claim using the key. Hence, Claims are
added to the ACS under the authority of a key.

> Each entry in the `measurement-values-map` is Evidence or an Endorsement

Reviewer 2: Please create an issue as to how Endorsement matching instead of Evidence is relevant here.

Author: Can you file the issue? I am not sure what you would like to see changed.

> Each entry in the `measurement-values-map` is Evidence or an Endorsement

Author: We haven't defined EMT yet (that is in the PR #145). I think part of the purpose of this paragraph has been replaced by the introduction 2 paras above. This one and the next are talking about multiple measurements in the same environment etc. I will rephrase.

describing a measurement associated with the environment identified in
the `environment-map`.

An Attester can provide multiple `environment-map`s each containing a
`measurement-values-map` with one entry; a single `environment-map` containing
multiple entries in its `measurement-values-map`; or a combination of
these approaches.

> There can be multiple entries in `evidence-triples` which have the

Author: I think this text should go elsewhere. It is not describing ACS

same `environment-map` and a different `authorized-by` field (see {{sec-authorized-by}}).

If Evidence or Endorsements from different sources has the same `environment-map`

> and `authorized-by` then the `measurement-values-map`s are merged.

Author: Text added

measurement values are not equivalent then the verifier SHALL report
an error and stop validation processing.

> #### Accepted Claims Set Initialization

Author: ?

Evidence formats may require format translation before being added to the Accepted Claims Set.
If format translation is required, a CoRIM profile, see {{sec-corim-profile-types}}, defines an Evidence translation function.

{{sec-dice-spdm}} provides information on how DICE and SPDM Evidence is reformatted into CoMID schema compliant expressions before being added to the Accepted Claims Set.

#### The authorized-by field in Accepted Claims Set {#sec-authorized-by}

The `authorized-by` field in an Accepted Claims Set entry indicates the entity
whose authority backs the claim.
An attacker may add entries to the Accepted Claims Set which are outside its area
of competence. Correct processing of the `authorized-by` field lets the Verifier
filter out claims from Attesters or Endorsers which do not meet its policies.

When adding cryptograpically signed Evidence to the Accepted Claims Set, the
Verifier SHALL set the `authorized-by` field in that Evidence to the trusted
authority keys at the head of each key chain which signed that Evidence. This
key is often the subject of a self-signed certificate.

If multiple authorities approve the Evidence, for example if multiple key chains

> are available, then the `authorized-by` field SHOULD include the key used by

Reviewer 3: ```suggestion
are available, then the ACS `authorized-by` field MUST be updated to include the key used by
```

each of those authorities.

When adding Endorsements resulting from CoRIM processing (see {{sec-add-to-acs}})

> the Verifier SHALL set the `authorized-by` field in that evidence to the

Reviewer 2: ```suggestion
the Verifier SHALL set the `authorized-by` field in that Evidence to the
```

{{sec-dice-spdm}} provides information on how DICE and SPDM Evidence is reformatted into CoMID schema compliant expressions before being added to the Accepted Claims Set.

#### The authorized-by field in Accepted Claims Set {#sec-authorized-by}

The `authorized-by` field in an Accepted Claims Set entry indicates the entity
whose authority backs the claim.

> An attacker may add entries to the Accepted Claims Set which are outside its area

Author: Thanks for the improved text. Replaced.

of competence. Correct processing of the `authorized-by` field lets the Verifier
filter out claims from Attesters or Endorsers which do not meet its policies.

When adding cryptograpically signed Evidence to the Accepted Claims Set, the
Verifier SHALL set the `authorized-by` field in that evidence to the trusted

> authority keys at the head of each key chain which signed that Evidence. This

Author: I think we need to be precise about what is an authority. Can you provide alternative text which tells the verifier exactly what to do?

If any part of the processing indicates that the Reference Value does not match then the remaining steps in this section are skipped for that group.

A Reference Value consists of an `environment-map` plus a `measurement-map`. In the
`reference-triple-record` these are encoded together. In other triples multiple
Reference Values are represented more compactly by letting one `environment-map`
apply to multiple `measurement-map`s.

The Verifier first looks for entries in the Accepted Claims Set with the same
`environment-map` as the Reference Value. These are the candidate claims. If there are
no candidate claims then the Reference Value does not match.

A Verifier SHALL compare two `environment-map`s using a binary comparison of the CBOR
encoded objects.

A Verifier SHOULD convert `environment-map` into a form which meets CBOR Core
Deterministic Encoding Requirements {{-cbor}} before performing the binary comparison.

> If the Reference Value contains an `authorized-by` field then the Verifier
> SHALL discard any candidate claims whose `authorized-by` field does
> not contain one of the keys listed in the Reference Value `authorized-by`

Reviewer 3: ```suggestion
If the Reference Value claims under the authority of the RVP match claims under the authority of the Attester then the Verifier
SHALL track these claims in the ACS under the authority of both asserters.
If the Verifier has an exclusive appraisal policy that specifies an authority to trust for these claims, the untrusted authorities SHOULD be removed.
If removal of an authority results in claims having no authority, the claims SHOULD be removed.
See the `authorized-by`
```

> field (see {{sec-authorized-by}} for more details).

Reviewer 3: ```suggestion
field {{sec-authorized-by}} for more details.
```

If any part of the processing indicates that the Reference Value does not match then the remaining steps in this section are skipped for that group.

A Reference Value consists of an `environment-map` plus a `measurement-map`. In the
`reference-triple-record` these are encoded together. In other triples multiple
Reference Values are represented more compactly by letting one `environment-map`
apply to multiple `measurement-map`s.

The Verifier first looks for entries in the Accepted Claims Set with the same

> `environment-map` as the Reference Value. These are the candidate claims. If there are
> no candidate claims then the Reference Value does not match.

Reviewer 2: ```suggestion
`environment-map` as the Reference Value. These are the candidate claims. If there are no 
candidate claims then the Reference Value does not match.
```
